reactos/drivers/usb/usbohci_new/hardware.h

```diff
@@ -0,0 +1,313 @@
+#define OHCI_NUMBER_OF_INTERRUPTS    32
+#define OHCI_MAX_PORT_COUNT          15
+#define ED_EOF                       -1
+#define OHCI_MAXIMUM_OVERHEAD        210 // 5.4  FrameInterval Counter, in bit-times
+#define OHCI_DEFAULT_FRAME_INTERVAL  11999 // 6.3.1  Frame Timing
+#define OHCI_MINIMAL_POTPGT          25 // == 50 ms., PowerOnToPowerGoodTime (HcRhDescriptorA Register)
+
+/* Controller states */
+#define OHCI_HC_STATE_RESET       0
+#define OHCI_HC_STATE_RESUME      1
+#define OHCI_HC_STATE_OPERATIONAL 2
+#define OHCI_HC_STATE_SUSPEND     3
+
+/* Endpoint Descriptor Control */
+#define OHCI_ED_DATA_FLOW_DIRECTION_FROM_TD 0
+#define OHCI_ED_DATA_FLOW_DIRECTION_OUT     1
+#define OHCI_ED_DATA_FLOW_DIRECTION_IN      2
+
+#define OHCI_ENDPOINT_FULL_SPEED 0
+#define OHCI_ENDPOINT_LOW_SPEED  1
+
+#define OHCI_ENDPOINT_GENERAL_FORMAT     0
+#define OHCI_ENDPOINT_ISOCHRONOUS_FORMAT 1
+
+/* Transfer Descriptor Control */
+#define OHCI_TD_INTERRUPT_IMMEDIATE 0
+#define OHCI_TD_INTERRUPT_NONE      7
+
+#define OHCI_TD_DIRECTION_PID_SETUP    0
+#define OHCI_TD_DIRECTION_PID_OUT      1
+#define OHCI_TD_DIRECTION_PID_IN       2
+#define OHCI_TD_DIRECTION_PID_RESERVED 3
+
+#define OHCI_TD_DATA_TOGGLE_FROM_ED  0
+#define OHCI_TD_DATA_TOGGLE_DATA0    2
+#define OHCI_TD_DATA_TOGGLE_DATA1    3
+
+#define OHCI_TD_CONDITION_NO_ERROR          0x00
+#define OHCI_TD_CONDITION_CRC_ERROR         0x01
+#define OHCI_TD_CONDITION_BIT_STUFFING      0x02
+#define OHCI_TD_CONDITION_TOGGLE_MISMATCH   0x03
+#define OHCI_TD_CONDITION_STALL             0x04
+#define OHCI_TD_CONDITION_NO_RESPONSE       0x05
+#define OHCI_TD_CONDITION_PID_CHECK_FAILURE 0x06
+#define OHCI_TD_CONDITION_UNEXPECTED_PID    0x07
+#define OHCI_TD_CONDITION_DATA_OVERRUN      0x08
+#define OHCI_TD_CONDITION_DATA_UNDERRUN     0x09
+#define OHCI_TD_CONDITION_BUFFER_OVERRUN    0x0C
+#define OHCI_TD_CONDITION_BUFFER_UNDERRUN   0x0D
+#define OHCI_TD_CONDITION_NOT_ACCESSED      0x0E
+
+typedef union _OHCI_TRANSFER_CONTROL {
+  struct {
+    ULONG Reserved       : 18;
+    ULONG BufferRounding : 1;
+    ULONG DirectionPID   : 2;
+    ULONG DelayInterrupt : 3;
+    ULONG DataToggle     : 2;
+    ULONG ErrorCount     : 2;
+    ULONG ConditionCode  : 4;
+  };
+  ULONG  AsULONG;
+} OHCI_TRANSFER_CONTROL, *POHCI_TRANSFER_CONTROL;
+
+typedef struct _OHCI_TRANSFER_DESCRIPTOR { // must be aligned to a 16-byte boundary
+  OHCI_TRANSFER_CONTROL Control;
+  PVOID CurrentBuffer; // physical address of the next memory location
+  ULONG_PTR * NextTD; // pointer to the next TD on the list of TDs
+  PVOID BufferEnd; // physical address of the last byte
+} OHCI_TRANSFER_DESCRIPTOR, *POHCI_TRANSFER_DESCRIPTOR;
+
+C_ASSERT(sizeof(OHCI_TRANSFER_DESCRIPTOR) == 16);
```

Comment:
This structure contains 32-bit fields:
4.3.1  General Transfer Descriptor
4.3.1.1  General Transfer Descriptor Format
What types do you propose to use instead of PVOID and PULONG_PTR?

Reply:
If a field is defined to be 32 bits by the specification, ULONG should do the right thing.

Reply:
Ok, replaced on ULONG.